unittesting/core/st3/fast_runner.py

```diff
@@ -0,0 +1,167 @@
+    # Rely on late binding in case a user patches it
+    sublime.set_timeout(partial(callback, *args, **kwargs), delay)
+
+
+DEFAULT_CONDITION_POLL_TIME = 17
+DEFAULT_CONDITION_TIMEOUT = 4000
+AWAIT_WORKER = 'AWAIT_WORKER'
+# Extract `set_timeout_async`, t.i. *avoid* late binding, in case a user
+# patches it
+run_on_worker = sublime.set_timeout_async
+
+
+class FastDeferringTextTestRunner(TextTestRunner):
+    """This test runner runs tests in deferred slices."""
+
+    def run(self, test):
+        """Run the given test case or test suite."""
+        self.finished = False
+        result = self._makeResult()
+        registerResult(result)
+        result.failfast = self.failfast
+        result.buffer = self.buffer
+        startTime = time.time()
+
+        def _start_testing():
+            with warnings.catch_warnings():
+                if self.warnings:
+                    # if self.warnings is set, use it to filter all the warnings
+                    warnings.simplefilter(self.warnings)
+                    # if the filter is 'default' or 'always', special-case the
+                    # warnings from the deprecated unittest methods to show them
+                    # no more than once per module, because they can be fairly
+                    # noisy.  The -Wd and -Wa flags can be used to bypass this
+                    # only when self.warnings is None.
+                    if self.warnings in ['default', 'always']:
+                        warnings.filterwarnings(
+                            'module',
+                            category=DeprecationWarning,
+                            message='Please use assert\\w+ instead.')
+                startTestRun = getattr(result, 'startTestRun', None)
+                if startTestRun is not None:
+                    startTestRun()
+                try:
+                    deferred = test(result)
+                    _continue_testing(deferred)
+
+                except Exception as e:
+                    _handle_error(e)
+
+        def _continue_testing(deferred, send_value=None, throw_value=None):
+            try:
+                if throw_value:
+                    condition = deferred.throw(throw_value)
+                else:
+                    condition = deferred.send(send_value)
+
+                if callable(condition):
+                    defer(0, _wait_condition, deferred, condition)
+                elif isinstance(condition, dict) and "condition" in condition and \
+                        callable(condition["condition"]):
+                    period = condition.get("period", DEFAULT_CONDITION_POLL_TIME)
+                    defer(period, _wait_condition, deferred, **condition)
+                elif isinstance(condition, int):
+                    defer(condition, _continue_testing, deferred)
+                elif condition == AWAIT_WORKER:
+                    run_on_worker(
+                        partial(defer, 0, _continue_testing, deferred)
```

Comment:
Once a TestCase yields to AWAIT_WORKER, the whole test case will be executed in the worker thread.
Should we send the coroutine back to the main thread like this?

```
partial(defer, 0, lambda: sublime.set_timeout(_continue_testing), deferred)
```

Reply:
I already do the double-hop here to get it back on the main thread.

This line evaluates to

```
timeout_async(lambda: timeout(lambda: continue))
^^^ await worker      ^^^ hop back    ^^^actually continue
```

but this quickly gets confusing, so we may need a test here

Reply:
Thanks. I missed the point that `defer` is really `set_timeout`.